File: VercelStatus/Sources/VercelStatus/Models/Deployment.swift
import Foundation

enum DeploymentState: String, Codable, Sendable, CaseIterable {
    case queued = "QUEUED"
    case initializing = "INITIALIZING"
    case building = "BUILDING"
    case ready = "READY"
    case error = "ERROR"
    case canceled = "CANCELED"
    case blocked = "BLOCKED"

    var isInProgress: Bool {
        switch self {
        case .queued, .initializing, .building:
            true
        default:
            false
        }
    }

    var isTerminal: Bool {
        switch self {
        case .ready, .error, .canceled, .blocked:
            true
        default:
            false
        }
    }

    var displayName: String {
        switch self {
        case .queued: "Queued"
        case .initializing: "Initializing"
        case .building: "Building"
        case .ready: "Ready"
        case .error: "Error"
        case .canceled: "Canceled"
        case .blocked: "Blocked"
        }
    }

    var shortLabel: String {
        switch self {
        case .queued: "Q"
        case .initializing: "INIT"
        case .building: "BUILD"
        case .ready: "OK"
        case .error: "ERR"
        case .canceled: "CAN"
        case .blocked: "BLK"
        }
    }
}

struct DeploymentMeta: Codable, Sendable {
    let githubCommitMessage: String?
    let githubCommitRef: String?
    let githubCommitSha: String?
    let githubCommitAuthorName: String?
}

struct DeploymentCreator: Codable, Sendable {
    let username: String?
}

struct Deployment: Codable, Identifiable, Sendable {
    let uid: String
    let name: String
    let url: String?
    let state: DeploymentState?
    let readyState: DeploymentState?
    let target: String?
    let createdAt: TimeInterval?
    let buildingAt: TimeInterval?
    let ready: TimeInterval?
    let meta: DeploymentMeta?
    let creator: DeploymentCreator?

    var id: String { uid }

    var status: DeploymentState {
        readyState ?? state ?? .queued
    }

    var previewURL: URL? {
        guard let url, !url.isEmpty else { return nil }
        if url.hasPrefix("http") {
            return URL(string: url)
        }
        return URL(string: "https://\(url)")
    }

    var branch: String? {
        meta?.githubCommitRef
    }

    var commitMessage: String? {
        meta?.githubCommitMessage
    }

    var commitSHA: String? {
        meta?.githubCommitSha.map { String($0.prefix(7)) }
    }

    var isProduction: Bool {
        target?.lowercased() == "production"
    }

    var createdDate: Date? {
        guard let createdAt else { return nil }
        return Date(timeIntervalSince1970: createdAt / 1000)
    }

    var readyDate: Date? {
        guard let ready else { return nil }
        return Date(timeIntervalSince1970: ready / 1000)
    }

    /// Best available timestamp for when the deployment last completed or started.
    var deployedDate: Date? {
        readyDate ?? createdDate
    }

    var durationSeconds: TimeInterval? {
        guard let buildingAt else { return nil }
        let end = ready ?? Date().timeIntervalSince1970 * 1000
        return (end - buildingAt) / 1000
    }
}

struct DeploymentsResponse: Codable, Sendable {
    let deployments: [Deployment]
}

struct DeploymentDetailResponse: Codable, Sendable {
    let uid: String?
    let name: String?
    let url: String?
    let state: DeploymentState?
    let readyState: DeploymentState?
    let target: String?
    let createdAt: TimeInterval?
    let buildingAt: TimeInterval?
    let ready: TimeInterval?
    let meta: DeploymentMeta?
    let creator: DeploymentCreator?

    func asDeployment(fallbackID: String) -> Deployment {
        Deployment(
            uid: uid ?? fallbackID,
            name: name ?? "deployment",
            url: url,
            state: state,
            readyState: readyState,
            target: target,
            createdAt: createdAt,
            buildingAt: buildingAt,
            ready: ready,
            meta: meta,
            creator: creator
        )
    }
}

File: toast-win/src/Toast.Core/Models/Deployment.cs
using System.Text.Json.Serialization;

namespace Toast.Core.Models;

public enum DeploymentState
{
    Queued,
    Initializing,
    Building,
    Ready,
    Error,
    Canceled,
    Blocked,
}

public static class DeploymentStateExtensions
{
    public static bool IsInProgress(this DeploymentState state) =>
        state is DeploymentState.Queued or DeploymentState.Initializing or DeploymentState.Building;

    public static bool IsTerminal(this DeploymentState state) =>
        state is DeploymentState.Ready or DeploymentState.Error or DeploymentState.Canceled or DeploymentState.Blocked;

    public static string DisplayName(this DeploymentState state) => state switch
    {
        DeploymentState.Queued => "Queued",
        DeploymentState.Initializing => "Initializing",
        DeploymentState.Building => "Building",
        DeploymentState.Ready => "Ready",
        DeploymentState.Error => "Error",
        DeploymentState.Canceled => "Canceled",
        DeploymentState.Blocked => "Blocked",
        _ => state.ToString(),
    };

    public static DeploymentState Parse(string? value) => value?.ToUpperInvariant() switch
    {
        "QUEUED" => DeploymentState.Queued,
        "INITIALIZING" => DeploymentState.Initializing,
        "BUILDING" => DeploymentState.Building,
        "READY" => DeploymentState.Ready,
        "ERROR" => DeploymentState.Error,
        "CANCELED" or "CANCELLED" => DeploymentState.Canceled,
        "BLOCKED" => DeploymentState.Blocked,
        _ => DeploymentState.Queued,
    };
}

public sealed class DeploymentMeta
{
    [JsonPropertyName("githubCommitMessage")]
    public string? GithubCommitMessage { get; set; }

    [JsonPropertyName("githubCommitRef")]
    public string? GithubCommitRef { get; set; }

    [JsonPropertyName("githubCommitSha")]
    public string? GithubCommitSha { get; set; }

    [JsonPropertyName("githubCommitAuthorName")]
    public string? GithubCommitAuthorName { get; set; }
}

public sealed class DeploymentCreator
{
    [JsonPropertyName("username")]
    public string? Username { get; set; }
}

public sealed class Deployment
{
    [JsonPropertyName("uid")]
    public string Uid { get; set; } = "";

    [JsonPropertyName("name")]
    public string Name { get; set; } = "";

    [JsonPropertyName("url")]
    public string? Url { get; set; }

    [JsonPropertyName("state")]
    public string? StateRaw { get; set; }

    [JsonPropertyName("readyState")]
    public string? ReadyStateRaw { get; set; }

    [JsonPropertyName("target")]
    public string? Target { get; set; }

    [JsonPropertyName("createdAt")]
    public double? CreatedAt { get; set; }

    [JsonPropertyName("buildingAt")]
    public double? BuildingAt { get; set; }

    [JsonPropertyName("ready")]
    public double? Ready { get; set; }

    [JsonPropertyName("meta")]
    public DeploymentMeta? Meta { get; set; }

    [JsonPropertyName("creator")]
    public DeploymentCreator? Creator { get; set; }

    public DeploymentState Status =>
        DeploymentStateExtensions.Parse(ReadyStateRaw ?? StateRaw);

    public string? CommitMessage => Meta?.GithubCommitMessage;

    public string? Branch => Meta?.GithubCommitRef;

    public string? CommitSha => Meta?.GithubCommitSha is { Length: > 0 } sha
        ? sha[..Math.Min(7, sha.Length)]
        : null;

    public bool IsProduction =>
        string.Equals(Target, "production", StringComparison.OrdinalIgnoreCase);

    public Uri? PreviewUrl
    {
        get
        {
            if (string.IsNullOrWhiteSpace(Url))
            {
                return null;
            }

            if (Url.StartsWith("http", StringComparison.OrdinalIgnoreCase))
            {
                return Uri.TryCreate(Url, UriKind.Absolute, out var absolute) ? absolute : null;
            }

            return Uri.TryCreate($"https://{Url}", UriKind.Absolute, out var https) ? https : null;
        }
    }
}

public sealed class DeploymentsResponse
{
    [JsonPropertyName("deployments")]
    public List<Deployment> Deployments { get; set; } = [];
}
